homeassistant/components/tado/water_heater.py

```diff
@@ -0,0 +1,303 @@
+            self.async_schedule_update_ha_state(True)
+
+        async_dispatcher_connect(
+            self.hass,
+            SIGNAL_TADO_UPDATE_RECEIVED.format("zone", self.zone_id),
+            async_update_callback,
+        )
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return self._supported_features
+
+    @property
+    def name(self):
+        """Return the name of the entity."""
+        return self.zone_name
+
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
+
+    @property
+    def current_operation(self):
+        """Return current readable operation mode."""
+        return WATER_HEATER_MAP_TADO.get(self._current_operation)
+
+    @property
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        return self._target_temp
+
+    @property
+    def is_away_mode_on(self):
+        """Return true if away mode is on."""
+        return self._is_away
+
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes (readable)."""
+        return [MODE_AUTO, MODE_HEAT, MODE_OFF]
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement used by the platform."""
+        return TEMP_CELSIUS
+
+    @property
+    def min_temp(self):
+        """Return the minimum temperature."""
+        return self._min_temperature
+
+    @property
+    def max_temp(self):
+        """Return the maximum temperature."""
+        return self._max_temperature
+
+    def set_operation_mode(self, operation_mode):
+        """Set new operation mode."""
+        mode = None
+
+        if operation_mode == MODE_OFF:
+            mode = CONST_MODE_OFF
+        elif operation_mode == MODE_AUTO:
+            mode = CONST_MODE_SMART_SCHEDULE
+        elif operation_mode == MODE_HEAT:
+            mode = self._default_overlay
+
+        self._current_operation = mode
+        self._overlay_mode = None
+
+        # Set a target temperature if we don't have any
+        if mode == CONST_OVERLAY_TADO_MODE and self._target_temp is None:
+            self._target_temp = self.min_temp
+            self.schedule_update_ha_state()
+
+        self._control_heater()
+
+    def set_temperature(self, **kwargs):
+        """Set new target temperature."""
+        temperature = kwargs.get(ATTR_TEMPERATURE)
+        if not self._supports_temperature_control or temperature is None:
+            return
+
+        self._current_operation = self._default_overlay
+        self._overlay_mode = None
+        self._target_temp = temperature
+        self._control_heater()
+
+    def update(self):
+        """Handle update callbacks."""
+        _LOGGER.debug("Updating water_heater platform for zone %d", self.zone_id)
+        try:
+            data = self._tado.data["zone"][self.zone_id]
+        except KeyError:
```

Comment:
When can this happen?

Reply:
This used to happen before, when the platforms were polling, and the main component had no data yet. This should not happen anymore. I've also removed it in `climate.py`.